cng/dsa_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build windows
+// +build windows
+
+package cng_test
+
+import (
+	"crypto/dsa"
+	"math/big"
+	"testing"
+
+	"github.com/microsoft/go-crypto-winnative/cng"
+	"github.com/microsoft/go-crypto-winnative/cng/bbig"
+)
+
+func TestGenerateDSAParameters(t *testing.T) {
+	testGenerateDSAParameters(t, 1024, 160)
+	testGenerateDSAParameters(t, 2048, 256)
+	testGenerateDSAParameters(t, 3072, 256)
+}
+
+func testGenerateDSAParameters(t *testing.T, L, N int) {
+	params, err := cng.GenerateDSAParameters(L)
+	if err != nil {
+		t.Errorf("%d-%d: error generating parameters: %s", L, N, err)
+		return
+	}
+
+	P := bbig.Dec(params.P)
+	Q := bbig.Dec(params.Q)
+	G := bbig.Dec(params.G)
+
+	if P.BitLen() != L {
+		t.Errorf("%d-%d: params.BitLen got:%d want:%d", L, N, P.BitLen(), L)
+	}
+
+	if Q.BitLen() != N {
+		t.Errorf("%d-%d: q.BitLen got:%d want:%d", L, N, Q.BitLen(), L)
+	}
+
+	one := new(big.Int)
+	one.SetInt64(1)
+	pm1 := new(big.Int).Sub(P, one)
+	quo, rem := new(big.Int).DivMod(pm1, Q, new(big.Int))
+	if rem.Sign() != 0 {
+		t.Errorf("%d-%d: p-1 mod q != 0", L, N)
+	}
+	x := new(big.Int).Exp(G, quo, P)
+	if x.Cmp(one) == 0 {
+		t.Errorf("%d-%d: invalid generator", L, N)
+	}
+
+	priv, err := cng.GenerateKeyDSA(params)
+	if err != nil {
+		t.Errorf("error generating key: %s", err)
+		return
+	}
+
+	testDSASignAndVerify(t, L, priv)
+}
+
+func testDSASignAndVerify(t *testing.T, i int, priv *cng.PrivateKeyDSA) {
+	hashed := []byte("testing")
+	r, s, err := cng.SignDSA(priv, hashed[:])
+	if err != nil {
+		t.Errorf("%d: error signing: %s", i, err)
+		return
+	}
+	params, X, Y, err := priv.Data()
+	if err != nil {
+		t.Errorf("%d: error exporting key: %s", i, err)
+		return
+	}
+	pub, err := cng.NewPublicKeyDSA(params, Y)
+	if err != nil {
+		t.Errorf("%d: error getting public key: %s", i, err)
+		return
+	}
+	if !cng.VerifyDSA(pub, hashed[:], r, s) {
+		t.Errorf("%d: error verifying", i)
+		return
+	}
+
+	// Test compatibility with crypto/dsa.
+	// TODO: This fails, sometimes...
```

Comment:
Issue for the todo? Does this mean it's flaky, or that sometimes specific values don't work?